### DISCUSSION TITLE
[Question] poetry add pathlib failed

### DISCUSSION THREAD

**Discussion by scheung38** [python-poetry/poetry#4830] [Q&A]
poetry add pathlib
Using version ^1.0.1 for pathlib

Updating dependencies
Resolving dependencies... (0.4s)

Package operations: 1 install, 0 updates, 0 removals

  • Installing pathlib (1.0.1): Failed

  EnvCommandError

  Command ['$HOME/Library/Caches/pypoetry/virtualenvs/devex-cli-oqezQ_Pz-py3.10/bin/pip', 'install', '--no-deps', 'file://$HOME/Library/Caches/pypoetry/artifacts/9c/d2/d7/58b5af0350587d8ebbcde29be9253a3f5445d25d3382ede7e4675995fc/pathlib-1.0.1.tar.gz'] errored with the following return code 1, and output: 
  Processing $HOME/Library/Caches/pypoetry/artifacts/9c/d2/d7/58b5af0350587d8ebbcde29be9253a3f5445d25d3382ede7e4675995fc/pathlib-1.0.1.tar.gz
    Preparing metadata (setup.py): started
    Preparing metadata (setup.py): finished with status 'error'
    ERROR: Command errored out with exit status 1:
     command: $HOME/Library/Caches/pypoetry/virtualenvs/devex-cli-oqezQ_Pz-py3.10/bin/python -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/04/cc5jymsj29s0d7_20tq44n4r0000gn/T/pip-req-build-vxckhzc3/setup.py'"'"'; __file__='"'"'/private/var/folders/04/cc5jymsj29s0d7_20tq44n4r0000gn/T/pip-req-build-vxckhzc3/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/var/folders/04/cc5jymsj29s0d7_20tq44n4r0000gn/T/pip-pip-egg-info-cdodwaz2
         cwd: /private/var/folders/04/cc5jymsj29s0d7_20tq44n4r0000gn/T/pip-req-build-vxckhzc3/
    Complete output (13 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "$HOME/Library/Caches/pypoetry/virtualenvs/devex-cli-oqezQ_Pz-py3.10/lib/python3.10/site-packages/setuptools/__init__.py", line 17, in <module>
        import setuptools.version
      File "$HOME/Library/Caches/pypoetry/virtualenvs/devex-cli-oqezQ_Pz-py3.10/lib/python3.10/site-packages/setuptools/version.py", line 1, in <module>
        import pkg_resources
      File "$HOME/Library/Caches/pypoetry/virtualenvs/devex-cli-oqezQ_Pz-py3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 23, in <module>
        import zipfile
      File "$HOME/.asdf/installs/python/3.10.0/lib/python3.10/zipfile.py", line 19, in <module>
        import pathlib
      File "/private/var/folders/04/cc5jymsj29s0d7_20tq44n4r0000gn/T/pip-req-build-vxckhzc3/pathlib.py", line 10, in <module>
        from collections import Sequence
    ImportError: cannot import name 'Sequence' from 'collections' (/Users/sebastiancheung/.asdf/installs/python/3.10.0/lib/python3.10/collections/__init__.py)
    ----------------------------------------
  WARNING: Discarding file:///Users/sebastiancheung/Library/Caches/pypoetry/artifacts/9c/d2/d7/58b5af0350587d8ebbcde29be9253a3f5445d25d3382ede7e4675995fc/pathlib-1.0.1.tar.gz. Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
  ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
  

  at ~/Library/Application Support/pypoetry/venv/lib/python3.10/site-packages/poetry/utils/env.py:1183 in _run
      1179│                 output = subprocess.check_output(
      1180│                     cmd, stderr=subprocess.STDOUT, **kwargs
      1181│                 )
      1182│         except CalledProcessError as e:
    → 1183│             raise EnvCommandError(e, input=input_)
      1184│ 
      1185│         return decode(output)
      1186│ 
      1187│     def execute(self, bin, *args, **kwargs):

- **OS version and name**:  Mac Monterey 12.0.1

- **Poetry version**: Poetry version 1.1.11

python --version
Python 3.10.0

**Comment by abn** [comment]
@scheung38 not sure why you would need `pathlib` as an external anymore unless you want to support `python < 3.4` in your project. For later python versions, [it is in the standard lib](https://docs.python.org/3/library/pathlib.html). The installation above fails because poetry falls back to source install if it cannot find a suitable wheel for your current intepreter version. To avoid this you can add a [python constraint](https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies).

```sh
poetry add pathlib --python "<3.4"
```

For those that are interested, the reason why poetry does not apply any python constraint by default, this is because the project metadata does not contain `python_requires`.

```console
$ curl -sL https://pypi.org/pypi/pathlib/json | jq -r '.info.python_requires'
null
```
